util/progress/progressui/display.go

```diff
@@ -159,6 +167,160 @@ func (v *vertex) update(c int) {
 	v.count += c
 }
 
+func (v *vertex) isStarted() bool {
+	return len(v.intervals) > 0
+}
+
+func (v *vertex) isCompleted() bool {
+	for _, ival := range v.intervals {
+		if ival.stop == nil {
+			return false
+		}
+	}
+	return true
+}
+
+type vertexGroup struct {
+	*vertex
+	subVtxs map[digest.Digest]client.Vertex
+}
+
+func (vg *vertexGroup) refresh() (changed, newlyStarted bool) {
+	newVtx := *vg.Vertex
+	newVtx.Cached = true
+	alreadyStarted := vg.isStarted()
+	for _, subVtx := range vg.subVtxs {
+		if subVtx.Started != nil {
+			newInterval := interval{
+				start: subVtx.Started,
+				stop:  subVtx.Completed,
+			}
+			prevInterval := vg.intervals[subVtx.Started.UnixNano()]
+			if !newInterval.isEqual(prevInterval) {
+				changed = true
+			}
+			if !alreadyStarted {
+				newlyStarted = true
+			}
+			vg.intervals[subVtx.Started.UnixNano()] = newInterval
+			if vg.mostRecentStart == nil || subVtx.Started.After(*vg.mostRecentStart) {
+				vg.mostRecentStart = subVtx.Started
+			}
+		}
+
+		// Group is considered cached iff all subvtxs are cached
+		newVtx.Cached = newVtx.Cached && subVtx.Cached
+
+		// Group error is set to the first error found in subvtxs, if any
+		if newVtx.Error == "" {
+			newVtx.Error = subVtx.Error
+		}
+	}
+
+	if vg.Cached != newVtx.Cached {
+		changed = true
+	}
+	if vg.Error != newVtx.Error {
+		changed = true
+	}
+	vg.Vertex = &newVtx
+
+	return changed, newlyStarted
+}
+
+type interval struct {
+	start *time.Time
+	stop  *time.Time
+}
+
+func (ival interval) duration() time.Duration {
+	if ival.start == nil {
+		return 0
+	}
+	if ival.stop == nil {
+		return time.Since(*ival.start)
+	}
+	return ival.stop.Sub(*ival.start)
+}
+
+func (ival interval) isEqual(other interval) (isEqual bool) {
+	return equalTimes(ival.start, other.start) && equalTimes(ival.stop, other.stop)
+}
+
+func equalTimes(t1, t2 *time.Time) bool {
+	if t2 == nil {
+		return t1 == nil
+	}
+	if t1 == nil {
+		return false
+	}
+	return t1.Equal(*t2)
+}
+
+// mergeIntervals takes a slice of (start, stop) pairs and returns a slice where
+// any intervals that overlap in time are combined into a single interval. If an
+// interval's stop time is nil, it is treated as positive infinity and consumes
+// any intervals after it. Intervals with nil start times are ignored and not
+// returned.
+func mergeIntervals(intervals []interval) []interval {
```

Comment:
Instead of these calculations wouldn't it be easier to just add the last `compled-started` diff every time vertex goes to completed state. If vertex is active then current time is `now-started+previous_sum`. Not a blocker but looks like it would be faster and simpler.

Reply:
Do you think the extra complexity might become an issue for this for big builds? Does this get called repeatedly after every 100ms on tty mode?

Reply:
It will get called for each `job` representing a vertex update (i.e. not status update, log, etc.) on each ui refresh. I just quickly ran a benchmark test using the `complex` test case in the unit tests I added and got 775 ns/op (in a linux VM running on an m1 macbook). That test case has 15 start/stop intervals to merge, whereas in the real world we'd usually expect at most 5. But round that up to 1 microsecond anyways, then assume that an extremely large build has 100 vertex `job`s per update and it still only takes 100 microseconds. Let me know if you think the estimate is off, but I'd be biased towards thinking the performance shouldn't make a difference here.